tests/models/test_models_unet_2d_condition.py

```diff
@@ -622,6 +675,36 @@ def test_asymmetrical_unet(self):
         # Check if input and output shapes are the same
         self.assertEqual(output.shape, expected_shape, "Input and output shapes do not match")
 
+    def test_ip_adapter(self):
+        init_dict, inputs_dict = self.prepare_init_args_and_inputs_for_common()
+
+        init_dict["attention_head_dim"] = (8, 16)
+
+        model = self.model_class(**init_dict)
+        model.to(torch_device)
+
+        with torch.no_grad():
+            sample1 = model(**inputs_dict).sample
+
+        ip_adapter_state_dict = create_ip_adapter_state_dict(model)
+        model._load_ip_adapter_weights(ip_adapter_state_dict)
+
+        assert model.config.encoder_hid_dim_type == "ip_image_proj"
+        assert model.encoder_hid_proj is not None
+
+        assert model.down_blocks[0].attentions[0].transformer_blocks[0].attn2.processor.__class__.__name__ in (
+            "IPAdapterAttnProcessor",
+            "IPAdapterAttnProcessor2_0",
+        )
+
+        batch_size = inputs_dict["encoder_hidden_states"].shape[0]
+        image_embeds = floats_tensor((batch_size, 1, model.cross_attention_dim)).to(torch_device)
+
+        with torch.no_grad():
+            sample2 = model(**{**inputs_dict, "added_cond_kwargs": {"image_embeds": image_embeds}}).sample
+
+        assert not sample1.allclose(sample2, atol=1e-4, rtol=1e-4)
```

Comment:
Can we make this test a bit stronger? E.g. we should also test the following:

- 1. having loaded and run forward once, let's replace the adapter with a different adapter and make sure results are different. Then load back the original one and make sure results are the same as the first one
- 2. Let's also add one test that checks for exact match

Reply:
done :)